nexus/src/db/datastore/update.rs

```diff
@@ -56,4 +64,168 @@ impl DataStore {
             .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
             .internal_context("deleting outdated available artifacts")
     }
+
+    pub async fn system_update_create(
+        &self,
+        opctx: &OpContext,
+        update: SystemUpdate,
+    ) -> CreateResult<SystemUpdate> {
+        // TODO: what's the right permission here?
+        opctx.authorize(authz::Action::CreateChild, &authz::FLEET).await?;
+
+        use db::schema::system_update::dsl::*;
+
+        diesel::insert_into(system_update)
+            .values(update.clone())
+            .on_conflict(version)
+            .do_nothing()
+            .returning(SystemUpdate::as_returning())
+            .get_result_async(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(
+                    e,
+                    ErrorHandler::Conflict(
+                        ResourceType::SystemUpdate,
+                        &update.version.to_string(),
+                    ),
+                )
+            })
+    }
+
+    pub async fn component_update_create(
+        &self,
+        opctx: &OpContext,
+        system_update_id: Uuid,
+        update: ComponentUpdate,
+    ) -> CreateResult<ComponentUpdate> {
+        // TODO: what's the right permission here?
+        opctx.authorize(authz::Action::CreateChild, &authz::FLEET).await?;
+
+        use db::schema::component_update;
+        use db::schema::system_update_component_update as join_table;
+
+        let result = diesel::insert_into(component_update::table)
+            .values(update.clone())
+            .on_conflict(component_update::columns::id) // TODO: should also conflict on version
+            .do_nothing()
+            .returning(ComponentUpdate::as_returning())
```

Comment:
See: https://github.com/oxidecomputer/omicron/issues/1168 - if this already existed, `result` would return `NotFound`

Reply:
Resolved in [6cdfe40](https://github.com/oxidecomputer/omicron/pull/2100/commits/6cdfe40954988d5e5c5e8d3dc33972b35df79075) by not doing `ON CONFLICT` at all. Because these are single row inserts, we can rely on unique constraints on id and version to produce the errors we want.